src/buffer/buffer_map.rs

```diff
@@ -53,57 +92,93 @@ impl BufferMap {
+    pub fn as_result(self) -> Result<(), Self> {
+        if !self.missing_buffers.is_empty() {
+            return Err(self);
+        }
+
+        if !self.incompatible_buffers.is_empty() {
+            return Err(self);
+        }
+
+        Ok(())
+    }
+
+    /// Same as [`Self::require_buffer_by_literal`], but can be used with
+    /// temporary borrows of a string slice. The string slice will be cloned if
+    /// an error message needs to be produced.
+    pub fn require_buffer_by_name<BufferType: 'static>(
         &mut self,
         expected_name: &str,
         buffers: &BufferMap,
-    ) -> Result<Buffer<Message>, ()> {
-        if let Some((name, buffer)) = buffers.inner.get_key_value(expected_name) {
-            if let Some(buffer) = buffer.downcast_for_message::<Message>() {
+    ) -> Result<BufferType, ()> {
+        let identifier = BufferIdentifier::Name(Cow::Borrowed(expected_name));
+        if let Some(buffer) = buffers.get(&identifier) {
+            if let Some(buffer) = buffer.downcast_buffer::<BufferType>() {
                 return Ok(buffer);
             } else {
                 self.incompatible_buffers.push(BufferIncompatibility {
-                    name: name.clone(),
-                    expected: std::any::type_name::<Buffer<Message>>(),
+                    identifier: BufferIdentifier::Name(Cow::Owned(expected_name.to_owned())),
+                    expected: std::any::type_name::<BufferType>(),
                     received: buffer.message_type_name(),
                 });
             }
         } else {
             self.missing_buffers
-                .push(Cow::Owned(expected_name.to_owned()));
+                .push(BufferIdentifier::Name(Cow::Owned(expected_name.to_owned())));
         }
 
         Err(())
     }
 
-    /// Check whether a named buffer is compatible with a specialized buffer type,
-    /// such as `JsonBuffer`.
-    pub fn require_buffer_type<BufferType: 'static>(
+    /// Check whether a named buffer is compatible with the required buffer type.
+    pub fn require_buffer_by_literal<BufferType: 'static>(
         &mut self,
-        expected_name: &str,
+        expected_name: &'static str,
+        buffers: &BufferMap,
+    ) -> Result<BufferType, ()> {
+        self.require_buffer::<BufferType>(BufferIdentifier::literal_name(expected_name), buffers)
+    }
+
+    /// Check whether an indexed buffer is compatible with the required buffer type.
+    pub fn require_buffer_by_index<BufferType: 'static>(
+        &mut self,
+        expected_index: usize,
+        buffers: &BufferMap,
+    ) -> Result<BufferType, ()> {
+        self.require_buffer::<BufferType>(BufferIdentifier::Index(expected_index), buffers)
+    }
+
+    /// Check whether the buffer associated with the identifier is compatible with
+    /// the required buffer type.
+    pub fn require_buffer<BufferType: 'static>(
+        &mut self,
+        identifier: BufferIdentifier<'static>,
```

Comment:
Could this take `impl Into<BufferIdentifier<'static>>` and support both name and index keys? Assuming we impl `From<usize> for BufferIdentifier<'static>`.

Reply:
I like this idea. I've changed it here: [530c92f](https://github.com/open-rmf/bevy_impulse/pull/52/commits/530c92f84eedcf012d588d6ff090b23ab9012dc9)

I changed the name to `require_buffer_by_identifier` in case in the future someone wants a more general `require_buffer` that could represent "This buffer type must exist somewhere in the map".